### DISCUSSION TITLE
how to set values received from api server?

### DISCUSSION THREAD

**Discussion by SlideAway** [react-hook-form/react-hook-form#6632] [Q&A]
Hi. I appreciate about making good library before getting to main question. 

It is helpful for me to make nice react project. 

Now, lets' get to the main question.

I have some problems about setting values and resetting values. 

Here's a simple code with codesandbox link
[code sandbox](https://codesandbox.io/s/react-hook-form-question1-9wr2k?file=/src/App.js)

```javascript
import "./styles.css";
import React, { useEffect, useState } from "react";
import { useForm, Controller } from "react-hook-form";

// api server send this data
const apiValue = {
  name: "james",
  nickname: "cage"
};

export default function App() {
  const { reset, control, getValues, setValue, handleSubmit } = useForm();
  const [data, setData] = useState(false);

  const save = (data) => {
    console.log(data);
  };

  useEffect(() => {
    reset({});
    if (data) {
      // Object.keys(apiValue).forEach((key) => setValue(key, apiValue[key]));

      reset({
        name: "james",
        nickname: "cage"
      });
    }
  }, [data]);

  return (
    <>
      data:{JSON.stringify(data)}
      <br />
      {JSON.stringify(getValues())}
      <div className="App">
        <Controller
          name="name"
          control={control}
          defaultValue=""
          as={<input />}
        />
      </div>
      <div className="App">
        <Controller
          name="nickname"
          control={control}
          defaultValue=""
          as={<input />}
        />
      </div>
      <button onClick={() => setData(true)}>set Values</button>
      <button onClick={() => setData(false)}>reset</button>
      <button onClick={handleSubmit(save)}>submit</button>
    </>
  );
}

```

I must set values sent by api server. 

data means it's adding or updating. 

if data is true, it's for updating, so i have to set the values by apiValue. 
if data is false, it's for adding, so i have to reset the form empty. 

I used both set value with forEach and reset. 

However, setValue doesn't work. 
reset({}) can set the values, but it doesn't reset the values. 

I know it's not the objective of  reset function. But I can't use react skillfully, and it's my best.

can you let me know how to set values properly?

it used react-hook-form v6

My english skill is not good, so if you cannot understand this question, can you comment to this question?

Thank you

**Comment by FPDK** [comment]
Hi,

If you don't call _reset_ first in your useEffect hook then it should work:
https://codesandbox.io/s/react-hook-form-question1-forked-1eou6?file=/src/App.js

Also regarding _getValues_, please see [the docs](https://react-hook-form.com/api/useform/getvalues):
An optimized helper for reading form values. The difference between watch and getValues is that getValues **will not trigger re-renders or subscribe to input changes**. 

If you instead use _useWatch_ then you can view your changed form object. In the sandbox link I have also added a _useWatch_ example.

**Reply by SlideAway** [reply]
Thank you for helping.

Maybe, I misunderstood about setValue. 

I think setValue is resolved thanks to FPDK

I'm sorry about lack of information about reset problem.   

if you click handlesubmit button after clicking set value, reset button in order, the values that have been reset restored 

it's displayed on console

do i use handlesubmit function wrong?
